desc/compute/utils.py

```diff
@@ -1204,6 +1204,112 @@ def surface_integrals_transform(grid, surface_label="rho"):
+        and are intended to assign more weight to some samples for reasons
+        unrelated to differences in uncertainty between samples. See
+        https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Reliability_weights.
+
+        The second case is when the weights are intended to assign more weight
+        to samples with less uncertainty. See
+        https://en.wikipedia.org/wiki/Inverse-variance_weighting.
+        The unbiased sample variance for this case is obtained by replacing the
+        effective number of samples in the formula this function implements,
+        :math:`n_e`, with the actual number of samples :math:`n`.
+
+        The third case is when the weights denote the integer frequency of each
+        sample. See
+        https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Frequency_weights.
+        This is indeed a distinct case from the above two because here the
+        weights encode additional information about the distribution.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to compute the sample variance.
+    weights : ndarray
+        Weight assigned to each sample of ``q``.
+        A good candidate for this parameter is the surface area Jacobian.
+    unbiased : bool
+        If this condition is set to false, then the biased estimator of the
+        sample variance is returned. This is desirable if you are only
+        concerned with computing the variance of the given set of numbers and
+        not the distribution the numbers are sampled from.
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute the variance over.
+    expand_out : bool
+        Whether to expand the output array so that the output has the same
+        shape as the input. Defaults to true so that the output may be
+        broadcast in the same way as the input. Setting to false will save
+        memory.
+
+    Returns
+    -------
+    variance : ndarray
+        Variance of the given weighted sample over each surface in the grid.
+        By default, the returned array has the same shape as the input.
+
+    """
+    _, _, spacing, _ = _get_grid_surface(grid, surface_label)
+    integrate = surface_integrals_map(grid, surface_label, expand_out=False)
+
+    v1 = integrate(weights)
+    v2 = integrate(weights**2 * jnp.prod(spacing, axis=-1))
+    # effective number of samples per surface
+    n_e = v1**2 / v2
+    # analogous to Bessel's bias correction
+    correction = n_e / (n_e - bool(unbiased))
+
+    q = jnp.atleast_1d(q)
+    # compute variance in two passes to avoid catastrophic round off error
+    mean = (integrate((weights * q.T).T).T / v1).T
+    mean = grid.expand(mean, surface_label)
+    variance = (correction * integrate((weights * ((q - mean) ** 2).T).T).T / v1).T
+    return grid.expand(variance, surface_label) if expand_out else variance
```

Comment:
All the `.T` you see are artifacts of how we vectorize the computation in the case `q` is a 2d or 3d array. The first axis of a multidimensional array iterates through the nodes of a grid. Numpy multiplies a 1d and 2d array by multiplying element wise against the last dimension. So we transpose the possible multidimensional array before a multiplication so that the last axis iterates through the nodes of a grid.

Reply:
When you compute the variance in two passes, what do you mean about catastrophic roundoff error if we only did one pass? I am just curious

Reply:
Catastrophic is the technical term https://en.wikipedia.org/wiki/Catastrophic_cancellation. The difference of squares example they give is what we want to avoid, as the squares will statistically be close. See http://www.cs.yale.edu/publications/techreports/tr222.pdf.